documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md

```diff
@@ -6,283 +6,268 @@ Proposal design for a platform invoke (P/Invoke) mechanism for Silk.NET 3.0.
 
-Custom code may be used as a call style by providing a pointer to SilkTouch using a Procedure Address Expression. This is useful in scenarios such as COM interop.
+Unless deemed inappropriate, inapplicable, and/or infeasible by the Silk.NET team (the team reserved the right to do so without Working Group approval), the types **should** have the following characteristics:
+- An instance **must** be constructable from a `ref T` or `ref readonly T` (depending on the outer dimension mutability) where `T` is either the generic type (or `byte` for non-generic variants), a "pointer-like" type of the inner dimension of type being constructed.
+- An indexer **must** be present accepting a `nuint` for the index, returning a `ref T` or `ref readonly T` (depending on the outer dimension mutability) where `T` is either the generic type (or `byte` for non-generic variants), a "pointer-like" type of the inner dimension of type being indexed.
+- A `GetPinnableReference` **must** be present taking no parameters and returning a `ref T` or `ref readonly T` (depending on the outer dimension mutability) where `T` is the native raw pointer representation of the inner dimension.
+- For single-dimension pointers, an `AsSpan` method **must** be present taking an `int` length argument (due to historical reasons in the .NET BCL) returning a `Span<T>` or `ReadOnlySpan<T>` (depending on the outer dimension mutability) where `T` the pointee type. For non-generic pointer-like types, a generic argument **may** be used to specify the type of the span.
+- An array of the same jagged dimension **must** be implicitly convertible to the type.
+- For single-dimension pointers, a multi-dimensional array (up to 3 dimensions) **must** be implicitly convertible to the type.
+- A raw pointer of the same dimensions **must** be implicitly convertible to the type.
+- `NullPtr` **must** be implicitly convertible to the type, and the returned pointer must represent a null reference.
+- An explicit operator **must** be present to unsafely convert the reference represented by the pointer to a raw pointer without the user using `fixed`/`GetPinnableReference`.
+- An explicit operator **must** be present to unsafely convert the reference represented by the pointer to a void pointer without the user using `fixed`/`GetPinnableReference`.
+- For generic pointer types, an implicit operator **must** be present to convert the pointer type to a `string` (or an array of strings of the inner dimension's jaggedness i.e. `PtrPtrPtr<byte>` becomes `string[][]`). Much like `ReadOnlySpan<char>`, `ToString` **must** call this implicitly if the type is a single-dimension pointer to a string pointee type. The implicit cast **may** throw an exception if the pointee type is not a string pointee type. This is because we can't constrain the type used on implicit operators on generic types.
```

Comment:
still feel uneasy with implicit string conversions for allocation/perf reasons but that would need benchmarking and further analysis and consideration

Reply:
If you don't want to marshal a string just don't pass a string 🤷

Reply:
Maxine brought this up in the discord, this may be better as an explicit operator as assuming a Ptr<T> equates to a string is dangerous.

Reply:
It is also probably good to limit this functionality to only String types (byte, sbyte, ushort, short, uint, int). Using extensions would work for a method, but we can't directly override an operator. It may be best to use an explicit function like ReadToString(), then the operator can call the function and we override that for specific types otherwise throw an exception.

Reply:
Agreed, I think we could define `ReadToString` in `SilkMarshal` as an extension method for `Ptr`, `Mut`, `Ptr<T>`, and `Mut<T>` where T is one of: byte, sbyte, ushort, short, char, uint, int